### PR TITLE
Implement custom Segment events

### PR DESCRIPTION
**Pull request by Mrtenz**
This implements custom events for Segment. Right now the following events are supported:

- `Snap Install Clicked`: When the user clicks on the "Add to MetaMask" button.
- `Snap Installed`: When a Snap is installed (or updated) without errors.
- `Snap Installation Failed`: When a Snap failed installation for an unknown reason.
- `Snap Installation Rejected`: When the user rejected the Snap installation.
- `Snap Update Clicked`: When the user clicks on the "Update Snap" button.
- `Snap Website Clicked`: When the user clicks on the Snap website (either in the installation modal, or on the Snap page).

### CODE REVIEW

File: src/utils/errors.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @param object - The object to check.
+ * @param property - The property to check for.
+ * @returns Whether the object has the property.
+ */
+export function hasProperty<Type extends object, Property extends string>(
+  object: Type,
+  property: Property,
+): object is Type & Record<Property, unknown> {
+  return Object.prototype.hasOwnProperty.call(object, property);
```
Comment (FrederikBolding) [inline src/utils/errors.ts:29]: `@metamask/utils`? 🤔

File: src/analytics.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  | SnapUpdateEvent;
+
+/**
+ * Track Snaps Directory events.
+ *
+ * @param event - The event to track.
+ */
+export function track(event: SnapEvent): void {
+  // TODO: Implement Segment.
+  console.log(event);
```
Comment (FrederikBolding) [inline src/analytics.ts:77]: I think it is as simple as `window.analytics.track()` once the plugin has run

Comment (Mrtenz) [reply src/analytics.ts:77]: It should be yeah. Just waiting until I can actually test this.

File: src/utils/errors.ts (hunk 1 above)
Comment (Mrtenz) [reply src/utils/errors.ts:29]: I can't get it to work without it breaking the build. 😞

Comment (FrederikBolding) [reply src/utils/errors.ts:29]: Can we leave a TODO about coming back to it?

**Review by FrederikBolding — DISMISSED** [review]

**Review by FrederikBolding — DISMISSED** [review]

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**